src/abstract_types.jl

```diff
@@ -0,0 +1,471 @@
+
+supports_trust_region(::AbstractDescentAlgorithm) = false
+supports_line_search(::AbstractDescentAlgorithm) = false
+
+get_linear_solver(alg::AbstractDescentAlgorithm) = __getproperty(alg, Val(:linsolve))
+
+"""
+    AbstractDescentCache
+
+Abstract Type for all Descent Caches.
+
+### `SciMLBase.solve!` specification
+
+```julia
+δu, success, intermediates = SciMLBase.solve!(cache::AbstractDescentCache, J, fu, u,
+    idx::Val; skip_solve::Bool = false, kwargs...)
+```
+
+  - `J`: Jacobian or Inverse Jacobian (if `pre_inverted = Val(true)`).
+  - `fu`: residual.
+  - `u`: current state.
+  - `idx`: index of the descent problem to solve and return. Defaults to `Val(1)`.
+  - `skip_solve`: Skip the direction computation and return the previous direction.
+    Defaults to `false`. This is useful for Trust Region Methods where the previous
+    direction was rejected and we want to try with a modified trust region.
+  - `kwargs`: keyword arguments to pass to the linear solver if there is one.
+
+#### Returned values
+
+  - `δu`: the descent direction.
+  - `success`: Certain Descent Algorithms can reject a descent direction for example
+    `GeodesicAcceleration`.
+  - `intermediates`: A named tuple containing intermediates computed during the solve.
+    For example, `GeodesicAcceleration` returns `NamedTuple{(:v, :a)}` containing the
+    "velocity" and "acceleration" terms.
+
+### Interface Functions
+
+  - `get_du(cache)`: get the descent direction.
+  - `get_du(cache, ::Val{N})`: get the `N`th descent direction.
+  - `set_du!(cache, δu)`: set the descent direction.
+  - `set_du!(cache, δu, ::Val{N})`: set the `N`th descent direction.
+  - `last_step_accepted(cache)`: whether or not the last step was accepted. Checks if the
+    cache has a `last_step_accepted` field and returns it if it does, else returns `true`.
+"""
+abstract type AbstractDescentCache end
+
+SciMLBase.get_du(cache::AbstractDescentCache) = cache.δu
+SciMLBase.get_du(cache::AbstractDescentCache, ::Val{1}) = get_du(cache)
+SciMLBase.get_du(cache::AbstractDescentCache, ::Val{N}) where {N} = cache.δus[N - 1]
+set_du!(cache::AbstractDescentCache, δu) = (cache.δu = δu)
+set_du!(cache::AbstractDescentCache, δu, ::Val{1}) = set_du!(cache, δu)
+set_du!(cache::AbstractDescentCache, δu, ::Val{N}) where {N} = (cache.δus[N - 1] = δu)
+
+function last_step_accepted(cache::AbstractDescentCache)
+    hasfield(typeof(cache), :last_step_accepted) && return cache.last_step_accepted
+    return true
+end
+
+"""
+    AbstractNonlinearSolveLineSearchAlgorithm
+
+Abstract Type for all Line Search Algorithms used in NonlinearSolve.jl.
+
+### `SciMLBase.init` specification
+
+```julia
+SciMLBase.init(prob::AbstractNonlinearProblem,
+    alg::AbstractNonlinearSolveLineSearchAlgorithm, f::F, fu, u, p, args...;
+    internalnorm::IN = DEFAULT_NORM,
+    kwargs...) where {F, IN} --> AbstractNonlinearSolveLineSearchCache
+```
+"""
+abstract type AbstractNonlinearSolveLineSearchAlgorithm end
```

Comment:
Wasn't this moved to LineSearch.jl?

Reply:
Not yet, but that is the eventual plan. We still need to figure out how to decouple the line searches like `RobustNonMonotoneLineSearch` which heavily ties into the cache.